art/attacks/evasion/graphite/graphite_whitebox_pytorch.py

```diff
@@ -0,0 +1,579 @@
+        while True:
+            adv_img = img.detach().clone()
+            adv_img = torch.where(mask_tensor > 0.5, prev_attack, adv_img)
+            rand_start = (
+                torch.FloatTensor(*img.size())
+                .uniform_(self.rand_start_epsilon_range[0], self.rand_start_epsilon_range[1])
+                .to(adv_img.device)
+            )
+            adv_img = (adv_img + mask_tensor * rand_start).detach()
+            adv_img = torch.clamp(adv_img, 0, 1).to(img.dtype)
+            adv_img.requires_grad = True
+            final_avg_grad = None
+
+            # Do EOT adversarial attack with current mask.
+            loop_length = self.steps if rounds > 0 else self.first_steps
+            for _ in range(loop_length):
+                avg_grad = torch.zeros(adv_img.size()).cuda()
+                for xform in xforms:
+                    xform_img = transform_wb(
+                        img.clone().unsqueeze(0),
+                        adv_img.unsqueeze(0),
+                        mask_tensor,
+                        xform,
+                        pts,
+                        self.net_size,
+                        clip_min,
+                        clip_max,
+                    )
+
+                    logits, _ = self.estimator._predict_framework(
+                        xform_img.cuda(), y_onehot_tensor
+                    )  # pylint: disable=W0212
+                    if self.use_logits:
+                        loss = torch.nn.functional.cross_entropy(
+                            input=logits,
+                            target=torch.tensor(target_label).unsqueeze(0).to(logits.device),
+                            reduction="mean",
+                        )
+                    else:
+                        loss = torch.nn.functional.nll_loss(
+                            input=logits, target=torch.tensor(target_label).unsqueeze(0), reduction="mean"
+                        )
+
+                    grad = torch.autograd.grad(loss, adv_img)[0]
+                    avg_grad += grad
+                avg_grad /= len(xforms)
+
+                avg_grad_sign = avg_grad.clone()
+                avg_grad_sign = torch.sign(avg_grad_sign)
+                avg_grad_sign[torch.isnan(avg_grad_sign)] = 0
+                adv_img = adv_img - mask_tensor * self.step_size * avg_grad_sign
+
+                adv_img = adv_img.clamp(0, 1)
+                transform_robustness = self._eval(
+                    img.detach().clone(),
+                    adv_img,
+                    mask_tensor,
+                    target_label,
+                    y_onehot_tensor,
+                    xforms,
+                    pts,
+                    clip_min,
+                    clip_max,
+                )
+                final_avg_grad = avg_grad
+                if transform_robustness >= self.min_tr:
+                    break
+
+            # make mypy happy
+            assert isinstance(transform_robustness, float)
```

Comment:
if you want to type check, you should use an if not isinstance(): raise SpecificError(errormsg).